lib/src/rules/prefer_contains.dart

```diff
@@ -0,0 +1,207 @@
+
+const useContains = 'Use contains instead of indexOf';
+const alwaysFalse =
+    'Always false because indexOf is always greater or equal -1.';
+const alwaysTrue = 'Always true because indexOf is always greater or equal -1.';
+
+const details = '''
+**DO NOT** use `.indexOf` to see if a collection contains an element.
+
+The `Iterable` contract does not require that a collection know its length or be
+able to provide it in constant time. Calling `.length` just to see if the
+collection contains anything can be painfully slow.
+
+Instead, there are faster and more readable getters: `.isEmpty` and
+`.isNotEmpty`. Use the one that doesn’t require you to negate the result.
+
+**GOOD:**
+```
+if (lunchBox.isEmpty) return 'so hungry...';
+if (words.isNotEmpty) return words.join(' ');
+```
+
+**BAD:**
+```
+if (lunchBox.length == 0) return 'so hungry...';
+if (words.length != 0) return words.join(' ');
+```
+''';
+
+class _LintCode extends LintCode {
+  static final registry = <String, LintCode>{};
+
+  factory _LintCode(String name, String message) => registry.putIfAbsent(
+      name + message, () => new _LintCode._(name, message));
+
+  _LintCode._(String name, String message) : super(name, message);
+}
+
+class PreferContains extends LintRule {
+  PreferContains()
+      : super(
+            name: 'prefer_contains',
+            description: desc,
+            details: details,
+            group: Group.style);
+
+  @override
+  AstVisitor getVisitor() => new Visitor(this);
+
+  void reportLintWithDescription(AstNode node, String description) {
+    if (node != null) {
+      reporter.reportErrorForNode(new _LintCode(name, description), node, []);
+    }
+  }
+}
+
+class Visitor extends SimpleAstVisitor {
+  final PreferContains rule;
+
+  Visitor(this.rule);
+
+  @override
+  visitSimpleIdentifier(SimpleIdentifier identifier) {
+    // Should be "indexOf".
+    Element propertyElement = identifier.bestElement;
+    if (propertyElement?.name != 'indexOf') {
+      return;
+    }
+
+    AstNode lengthAccess;
+    InterfaceType type;
+
+    AstNode parent = identifier.parent;
+    if (parent is MethodInvocation && identifier == parent.methodName) {
+      lengthAccess = parent;
+      if (parent.target?.bestType is! InterfaceType) {
+        return;
+      }
+      type = parent.target?.bestType;
+    } else {
+      return;
+    }
+
+    // Should be subtype of Iterable or String.
+    AnalysisContext context = propertyElement.context;
+    TypeProvider typeProvider = context.typeProvider;
+    TypeSystem typeSystem = context.typeSystem;
+
+    if (typeSystem.mostSpecificTypeArgument(type, typeProvider.iterableType) ==
+            null &&
+        !type.element.type.isSubtypeOf(typeProvider.stringType)) {
+      return;
+    }
+
+    AstNode search = lengthAccess;
+    while (
+        search != null && search is Expression && search is! BinaryExpression) {
+      search = search.parent;
+    }
+
+    if (search is! BinaryExpression) {
```

Comment:
Here trying to find binary comparison operator for .indexOf access. We should go up in AST because access could be wrapped in parentheses, or `as` expression.
Most of the time, it will only iterate once.

I'm not sure which part you want to extract into method.